controllers/configurationpolicy_controller.go

```diff
@@ -506,10 +506,30 @@ func (r *ConfigurationPolicyReconciler) getObjectTemplateDetails(
 	return templateObjs, selectedNamespaces, false, nil
 }
 
-func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy) []string {
+func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy,
+	newRelated []policyv1.RelatedObject,
+) []string {
 	deletionFailures := []string{}
+	objsToDelete := plc.Status.RelatedObjects
 
-	for _, object := range plc.Status.RelatedObjects {
+	// When spec is updated and new related objects are created
+	if len(newRelated) != 0 {
+		log.Info("preparing to delete detached objects...")
+
+		var objShouldRemoved []policyv1.RelatedObject
+
+		for _, oldR := range objsToDelete {
+			exist := containRelated(newRelated, oldR)
+
+			if !exist {
+				objShouldRemoved = append(objShouldRemoved, oldR)
+			}
+		}
+
+		objsToDelete = objShouldRemoved
```

Comment:
Optional simplification
```suggestion
		objsToDelete := []policyv1.RelatedObject{}

		for _, oldR := range objsToDelete {
			if !containRelated(newRelated, oldR) {
				objsToDelete = append(objsToDelete, oldR)
			}
		}
```